File: 2443-check-if-there-is-a-valid-partition-for-the-array/2443-check-if-there-is-a-valid-partition-for-the-array.cpp
class Solution {
public:
    bool helper(int ind,vector<int>& nums,vector<int>& dp){
        if(ind==nums.size()){
            return true;
        }
        if (dp[ind]!=-1){
            return dp[ind];
        }
        bool ans=false;
        if(ind+1<nums.size() && nums[ind]==nums[ind+1]){
            ans|=helper(ind+2,nums,dp);
        }
        if(ind+2<nums.size() && nums[ind]==nums[ind+1] && nums[ind+1]==nums[ind+2]){
            ans|=helper(ind+3,nums,dp);
        }
        else if(ind+2<nums.size() && nums[ind+2]==nums[ind+1]+1 && nums[ind+1]==nums[ind]+1){
            ans|=helper(ind+3,nums,dp);
        }
        
        return dp[ind]=ans;
    }
    bool validPartition(vector<int>& nums) {
        vector<int> dp(nums.size(),-1);
        return helper(0,nums,dp);
    }
};

File: 2443-check-if-there-is-a-valid-partition-for-the-array/2443-check-if-there-is-a-valid-partition-for-the-array.py
class Solution:
    def validPartition(self, nums: List[int]) -> bool:
        def calc(i):
            #print(i)
            if i>=len(nums):
                return True
            if dp[i]!=-1:
                return dp[i]
            two=False
            if i+1<len(nums):
                two=nums[i]==nums[i+1]
            thre1=False
            if i+2<len(nums):
                thre1=nums[i]==nums[i+1]==nums[i+2]
            thre=False
            if i+2<len(nums):
                thre=nums[i+2]==nums[i+1]+1==nums[i]+2
            #print(two,thre1,thre)
            dp[i]=(two and calc(i+2)) or (thre and calc(i+3)) or (thre1 and calc(i+3))
            return dp[i]
        dp=[-1 for i in range(len(nums))]
        return calc(0)